robot/http/client_driver.cpp

```diff
@@ -0,0 +1,140 @@
+#include "caf/actor_system_config.hpp"
+#include "caf/caf_main.hpp"
+#include "caf/detail/latch.hpp"
+#include "caf/ipv4_address.hpp"
+
+#include <iostream>
+#include <memory>
+#include <utility>
+
+using namespace std::literals;
+
+namespace http = caf::net::http;
+using namespace caf;
+
+// -- configuration setup ------------------------------------------------------
+struct config : actor_system_config {
+  config() {
+    opt_group{custom_options_, "global"}
+      .add<http::method>("method,m", "HTTP method to use")
+      .add<uri>("resource,r", "Requested resource (URL)")
+      .add<std::string>("payload,p", "Optional payload to send");
+  }
+};
+
+constexpr auto default_method = http::method::get;
+
+// -- http client sending get request and awaiting response --------------------
+class http_app : public http::upper_layer::client {
+public:
+  static auto make(http::method method, std::string resource,
+                   std::string payload) {
+    return std::make_unique<http_app>(method, std::move(resource),
+                                      std::move(payload));
+  }
+
+  http_app(http::method method, std::string resource, std::string payload)
+    : method_{method},
+      resource_{std::move(resource)},
+      payload_{std::move(payload)} {
+    latch_ = std::make_shared<detail::latch>(2);
+  }
+
+  ~http_app() {
+    latch_->count_down();
+  }
+
+  virtual void prepare_send() override {
+    // nop
+  }
+
+  [[nodiscard]] virtual bool done_sending() override {
+    return true;
+  }
+
+  virtual void abort(const caf::error& reason) override {
+    std::cerr << "*** abort called with error - " << to_string(reason)
+              << std::endl;
+    exit(EXIT_FAILURE);
+  }
+
+  caf::error start(http::lower_layer::client* ll) override {
+    down = ll;
+    // Send request.
+    down->begin_header(method_, resource_);
+    if (!payload_.empty())
+      down->add_header_field("Content-Length", std::to_string(payload_.size()));
+    down->end_header();
+    if (!payload_.empty())
+      down->send_payload(as_bytes(make_span(payload_)));
+    // Await response.
+    down->request_messages();
+    return caf::none;
+  }
+
+  ptrdiff_t consume(const http::response_header& hdr,
+                    caf::const_byte_span payload) override {
+    std::cout << hdr.version() << ' ' << hdr.status() << ' '
+              << hdr.status_text() << std::endl;
+    hdr.for_each_field([](const auto& key, const auto& val) {
+      if (key != "Date")
+        std::cout << key << ':' << val << std::endl;
+    });
+    std::cout << std::string_view{reinterpret_cast<const char*>(payload.data()),
+                                  payload.size()}
+              << std::endl;
+    return static_cast<ptrdiff_t>(payload.size());
+  }
+
+  std::shared_ptr<detail::latch> latch() const noexcept {
+    return latch_;
+  }
+
+private:
+  http::lower_layer::client* down = nullptr;
+  http::method method_;
+  std::string resource_;
```

Comment:
Got a bit confused about this member. The `resource` parameter for the program is an URI, but this captures only the path (plus query). So I think naming it `path_` would be clearer:

```suggestion
  std::string path_;
```